internal/state/store.go

```diff
@@ -575,30 +578,21 @@ func (store dbStore) SaveFinalizeBlockResponse(height int64, resp *abci.Finalize
 	}
 	resp.TxResults = dtxs
 
-	// If the flag is false then we save the ABCIResponse. This can be used for the /BlockResults
-	// query or to reindex an event using the command line.
-	if !store.DiscardABCIResponses {
-		bz, err := resp.Marshal()
-		if err != nil {
-			return err
-		}
-		if err := store.db.Set(calcABCIResponsesKey(height), bz); err != nil {
-			return err
-		}
-	}
-
-	// We always save the last ABCI response for crash recovery.
-	// This overwrites the previous saved ABCI Response.
-	response := &cmtstate.ABCIResponsesInfo{
-		FinalizeBlock: resp,
-		Height:        height,
-	}
-	bz, err := response.Marshal()
+	bz, err := resp.Marshal()
 	if err != nil {
 		return err
 	}
 
-	return store.db.SetSync(lastABCIResponseKey, bz)
+	// Save the ABCI response.
+	//
+	// We always save the last ABCI response for crash recovery.
+	// If `store.DiscardABCIResponses` is true, then we delete the previous ABCI response.
+	if store.DiscardABCIResponses && height > 1 {
+		if err := store.db.Delete(calcABCIResponsesKey(height - 1)); err != nil {
```

Comment:
While we do not double save a lot of data, we do have 2 DB operations here. Note that if compaction does not work as expected, this might still be blowing up `state.db` even though we are deleting data. 

If `DiscardABCIResponses` is true then we could potentially just save everything under `LastABCIResponsesKey`. 
So we have only one key in the DB and don't risk having potentially a big DB due to lack of compaction kicking in. 

The changes could look something like this: 

```
func  (store dbStore) LoadLastFinalizeBlockResponse(height int64) (*abci.FinalizeBlockResponse, error) {

.....
var bz
if store.DiscareABCIResponse == true {

    bz, err := store.db.Get(lastABCIResponseKey)
    
 } else {

    bz, err := store.db.Get(calcABCIResponseKey(height))
}
// Perform checks on height

....

```

Then in 

```
func (store dbStore) SaveFinalizeBlockResponse(height int64, resp*abci.FinalizeBlockResponse) error {

...

var resp 
if store.DiscardABCIResponses == true {
      store.db.SetSync(lastABCIResponseKey, bz)
} else {

  store.db.SetSync(calcLastABCIResponseKey(height), bz)
}
```

The only potential problem I see is if this setting (`DiscardABCIResponses`) can be changed between node boot-ups so there might be some additional checks to be added.

Reply:
> While we do not double save a lot of data, we do have 2 DB operations here. Note that if compaction does not work as expected, this might still be blowing up state.db even though we are deleting data.

Good point 👍 under what conditions compaction does not work as expected?

> The only potential problem I see is if this setting (DiscardABCIResponses) can be changed between node boot-ups so there might be some additional checks to be added.

This is a valid concern.

Reply:
> > While we do not double save a lot of data, we do have 2 DB operations here. Note that if compaction does not work as expected, this might still be blowing up state.db even though we are deleting data.
> 
> Good point 👍 under what conditions compaction does not work as expected?
> 

At the moment (without the changes we plan to introduce with in-process compaction, there is no guarantee that golevleDB will actually compact the storage, the same way it has not been doing so previously for blocks. . This is worth testing out  in this case though. If the problem exists, then calling `CompactRange` after a couple of these delete operations would solve the storage footprint problem. 

> > The only potential problem I see is if this setting (DiscardABCIResponses) can be changed between node boot-ups so there might be some additional checks to be added.
> 
> This is a valid concern.

I am not sure this is a problem in your version of the code, only in what I propose.

Reply:
@melekes , what could be worth running is a small testnet on this branch and monitor the storage usage. Even on one node.